x/emissions/module/abci.go

```diff
@@ -140,5 +125,30 @@ func EndBlocker(ctx context.Context, am AppModule) error {
 	}
 	wg.Wait()
 
+	// Close any open windows due this blockHeight
+	workerWindowsToClose := am.keeper.GetWorkerWindowTopicIds(sdkCtx, blockHeight)
+	for _, topicId := range workerWindowsToClose.TopicIds {
+		sdkCtx.Logger().Debug(fmt.Sprintf("ABCI EndBlocker: Worker close cadence met for topic: %d", topicId))
+		// Check if there is an unfulfilled nonce
+		nonces, err := am.keeper.GetUnfulfilledWorkerNonces(sdkCtx, topicId.TopicId)
+		if err != nil {
+			sdkCtx.Logger().Warn(fmt.Sprintf("Error getting unfulfilled worker nonces: %s", err.Error()))
+			continue
+		} else if len(nonces.Nonces) == 0 {
+			// No nonces to fulfill
+			continue
+		} else {
+			for _, nonce := range nonces.Nonces {
+				sdkCtx.Logger().Debug(fmt.Sprintf("ABCI EndBlocker: Closing Worker window for topic: %d, nonces: %v", topicId.TopicId, nonce))
+				err := allorautils.CloseWorkerNonce(&am.keeper, sdkCtx, topicId.TopicId, *nonce)
+				if err != nil {
+					sdkCtx.Logger().Warn(fmt.Sprintf("Error closing worker nonce: %s", err.Error()))
+					// Proactively close the nonce
+					am.keeper.FulfillWorkerNonce(sdkCtx, topicId.TopicId, nonce)
+				}
+			}
+		}
+	}
+	am.keeper.DeleteWorkerWindowBlockheight(sdkCtx, blockHeight)
```

Comment:
From logical perspective, this line should move to below line 149, isn't it? Because we do not add `workerwindow` on every block.

Reply:
nope, outside the loop. because it just removes the whole entry (if any)  after the iteration of the topics and the actions on it. 

An entry is essentially a (blockHeight, topicId[]), so once the topics are iterated, we remove the entry

Reply:
oh...yes. I looked at the loop incorrectly. What I wanted to do was to delete only if there was data for this blockHeight.

Reply:
hmmm right, this could be added for clarity actually.,

Reply:
done